[Core] Read properties independently

loumalouomega opened the pull request:
Fixes #6019 @thelfer

This PR creates an auxiliar Parameters when reading properties in order to avoid error, so these non-registered properties can be processed later

loumalouomega committed:
[Core] Read properties independently 

Fixes #6019 @thelfer

This PR creates an auxiliar Parameters when reading properties in order to avoid error, so these non-registered properties can be processed later

thelfer commented:
@loumalouomega Not sure that this is what I wanted, because the material properties passed to the CLs constructor is empty at this point !

Moreover, in this implementation you just discard unknonw material properties.

BTW is there a typo in the method name: shouldn't  `AssingMaterialToProperty` `AssignMaterialToProperty`. A brief search to "Assing" does not yield any thing be related to  what we are doing here :)

loumalouomega commented:
> 
> 
> @loumalouomega Not sure that this is what I wanted, because the material properties passed to the CLs constructor is empty at this point !
> 

True, this can be reordered without additional harm

> Moreover, in this implementation you just discard unknonw material properties.
> 
> BTW is there a typo in the method name: shouldn't `AssingMaterialToProperty` `AssignMaterialToProperty`. A brief search to "Assing" does not yield any thing be related to what we are doing here :)

Ups for the typo, not mine BTW

loumalouomega committed:
Suggestions

loumalouomega commented:
@thelfer now the CL is created after reading the variables

thelfer commented:
> @thelfer now the CL is created after reading the variables

ok but we lost the fact that for standard CLs unknown material properties should not be tolerated ! Maybe I have too stringent rules, but I personally would not allow it !

This is what I would do (just mock-up, I don't expect it to compile):

```c++
void ReadMaterialsUtility::AssingConstitutiveLawToProperty(const Parameters MaterialData,
                                                           Properties& rProperty) {
  KRATOS_TRY;
  // Set the CONSTITUTIVE_LAW for the current p_properties.
  if (MaterialData.Has("constitutive_law")) {
    Parameters cl_parameters = MaterialData["constitutive_law"];
    std::string constitutive_law_name = cl_parameters["name"].GetString();
    TrimComponentName(constitutive_law_name);
    cl_parameters["name"].SetString(constitutive_law_name);
    KRATOS_ERROR_IF_NOT(KratosComponents<ConstitutiveLaw>::Has(constitutive_law_name))
        << "Kratos components missing \"" << constitutive_law_name << "\"" << std::endl;
    auto variables = Parameters {}
    auto tables = Parameters {}
    if (MaterialData.Has("Variables")) {
      variables = MaterialData["Variables"];
    }
    if (MaterialData.Has("Tables")) {
      tables = MaterialData["Tables"];
    }
    auto p_constitutive_law = KratosComponents<ConstitutiveLaw>::Get(constitutive_law_name)
                                  .Create(cl_parameters, rProperty, variables, tables);
    rProperty.SetValue(CONSTITUTIVE_LAW, p_constitutive_law);
  } else {
    KRATOS_INFO("Read materials") << "No constitutive law defined for material ID: "
                                  << rProperty.Id() << std::endl;
    // Assign variables
    AssingVariablesToProperty(MaterialData, rProperty);
    // Assign tables
    AssingTablesToProperty(MaterialData, rProperty);
  }
  KRATOS_CATCH("");
}
```

And in the default version of the `Create` method, I would call  `AssignVariablesToPropery` and ` `AssignTablesToProperty` (which have to turned into static methods...)

RiccardoRossi reviewed (changes requested):
sorry guys but i think we are having a misunderstanding here...

as i understood we agreed to modify the Create so that the reading would be automatic and would be done through that function. Did i get it wrong?

thelfer commented:
@RiccardoRossi That the point of my mock-up: the `variables` (and `tables` which is an extension of the original proposal that could be discarded if it simplifies the discussion) are passed as argument to the `Create` method and treated there.

The default "new" `Create` method only populates  the properties and checks if the variables are registred *which is exactly the current behaviour*. 

This `Create` method is overriden in the `MGISApplication` so that variables expected by the `MFront` CLs  are treated. Unexpected variables shall then be treated as error.

I believe(d) that we agreed on this, didn't we ?

RiccardoRossi commented:
@thelfer your approach will not work since we cannot redefine the function Create of the CL...

the idea we understood was to have something like

1 - read variables and tables into properties
2 - call the create method - since we pass the "Properties",  that should already contain "Variables" and "Tables". (if it doesn't we can change the order at which Create and reading of variables is done)

this way there is no special reading_material_utility and everything happens in the create which is anyway user-specific.

thelfer commented:
@RiccardoRossi I may find that I am a bit stubborn (you would be right, btw), but I just implemented the previous code mockup. It does work as I expected and the `Create` of the `MGISConstitutiveLawFactory` is indeed called. This means that I do have complete control in this constructor on what the user has put in the `Variables` block. This is exactly what I wanted.

loumalouomega commented:
@RiccardoRossi, several CL override the create method (the factories for
example). I don't know what you meant with this

El jue., 19 dic. 2019 21:21, Riccardo Rossi <notifications@github.com>
escribió:

> @thelfer <https://github.com/thelfer> your approach will not work since
> we cannot redefine the function Create of the CL...
>
> the idea we understood was to have something like
>
> 1 - read variables and tables into properties
> 2 - call the create method - since we pass the "Properties", that should
> already contain "Variables" and "Tables". (if it doesn't we can change the
> order at which Create and reading of variables is done)
>
> this way there is no special reading_material_utility and everything
> happens in the create which is anyway user-specific.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/KratosMultiphysics/Kratos/pull/6136?email_source=notifications&email_token=AEYQZAH4EOECORMHX5BFBBDQZPJU7A5CNFSM4J5KFIK2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEHKZ66Q#issuecomment-567648122>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEYQZAG5NT7N2DIQSG33C6LQZPJU7ANCNFSM4J5KFIKQ>
> .
>

RiccardoRossi commented:
what i am telling is that you cannot modify the "read_material_utils.py" since it is used by everone ...

if you want to write a new one (possibly derived from the original) you will also need to derive your own solver.

Aside of this "Create" is a virtual function, so you need to use the API defined in the base class

RiccardoRossi commented:
@thelfer how are u calling that function from within the solver?
also, how are u calling your function "Create"? it is virtual, and the base class does not define the interface you are using...

loumalouomega commented:
@RiccardoRossi we modified read_materials_utility last week. If you look at
the changes this only ch, so it is not something impossible to change with
the proper discussion. This PR changes the error for a warning, does not
affect anything else (that's why all tests pass without further
modification). I added to the review to the technical committee because
even this small change affects an important file. This PR is also to
discuss, as we think we had a confusion about the way to proceed with the
material properties and the MGIS library.

El vie., 20 dic. 2019 8:42, Riccardo Rossi <notifications@github.com>
escribió:

> what i am telling is that you cannot modify the "read_material_utils.py"
> since it is used by everone ...
>
> if you want to write a new one (possibly derived from the original) you
> will also need to derive your own solver.
>
> Aside of this "Create" is a virtual function, so you need to use the API
> defined in the base class
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/KratosMultiphysics/Kratos/pull/6136?email_source=notifications&email_token=AEYQZACWHFUM2UH6QEVAO7DQZRZPJA5CNFSM4J5KFIK2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEHMFP7Y#issuecomment-567826431>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEYQZACKR5NFBQSVTRBA3JLQZRZPJANCNFSM4J5KFIKQ>
> .
>

thelfer commented:
Dear @RiccardoRossi, @loumalouomega, please find attach a patch that works for me (made it against the `mfront/application` branch). Again, it is just a proposal, but I believe its worth considering it.

I modified the `ReadMaterialsUtility::AssignMaterialToProperty` to extract the variables and tables parameters and pass them to the `Create` method of the consistitutive law. The default version uses them to fill the material properties. I just turned `ReadMaterialsUtility::AssignVariablesToProperty` and `ReadMaterialsUtility::AssignTablesToProperty` into free functions and call them in the default `Create` method. This is perfectly *backward compatible*

But now,  I am free to handle them as I want in the `Create` method of the `MGISConstitutiveLawFactory` class. This is exactly what I need.

I tested it, it seems to work fine. 

> what i am telling is that you cannot modify the "read_material_utils.py" since it is used by everone ...

> @thelfer how are u calling that function from within the solver?

No, The magic is in the `ReadMaterialsUtility::AssignConstitutiveLawToProperty` method which creates a CL and then call the `Create` method. Hence the virtual mechanism works as expected:

```c++
auto p_constitutive_law = KratosComponents<ConstitutiveLaw>::Get(constitutive_law_name)
                                  .Create(cl_parameters, variables, tables, rProperty);
```
[6019.zip](https://github.com/KratosMultiphysics/Kratos/files/3987303/6019.zip)

thelfer commented:
> This PR changes the error for a warning.

That is an important point that made me reluctant about @loumalouomega' proposal. 

Hence my proposal which does not have this caveat. It may have others (in particular, I don't know if the methods `ReadMaterialsUtility::AssignVariablesToProperty` and `ReadMaterialsUtility::AssignTablesToProperty` still have any meaning...

loumalouomega commented:
> 
> 
> > This PR changes the error for a warning.
> 
> That is an important point that made me reluctant about @loumalouomega' proposal.
> 
> Hence my proposal which does not have this caveat. It may have others (in particular, I don't know if the methods `ReadMaterialsUtility::AssignVariablesToProperty` and `ReadMaterialsUtility::AssignTablesToProperty` still have any meaning...

Can we summarize what should be changed? (aside of the typo which I am going to do an independent PR right now)

thelfer commented:
@loumalouomega In short:

- modifiy `ConstitutiveLaw::Create`' declaration as follows:
   ```c++
    virtual Pointer Create(Kratos::Parameters NewParameters,
                                      const Kratos::Parameters& variables,
                                      const Kratos::Parameters& tables,
                                      Properties& rProperties) const;
   ```
- modify `ConstitutiveLaw::Create`' implementation as follows:
   ```c++
   ConstitutiveLaw::Pointer ConstitutiveLaw::Create(Kratos::Parameters NewParameters,
                                                 const Kratos::Parameters& variables,
                                                 const Kratos::Parameters& tables,
                                                 Properties& rProperties) const {
     Kratos::AssignVariablesToProperty(variables, rProperties);
     Kratos::AssignTablesToProperty(tables, rProperties);
     return this->Create(NewParameters);
   } 
   ```
-  `ReadMaterialsUtility::AssignMaterialToProperty` to extract the variables and tables parameters and pass them to the Create method of the consistitutive law, as follows:
   ```c++
        auto variables = Parameters{};
        auto tables = Parameters{};
        if (MaterialData.Has("Variables")) {
          variables = MaterialData["Variables"];
        } else {
          KRATOS_INFO("Read materials")
              << "No variables defined for material ID: " << rProperty.Id() << std::endl;
        }
        if (MaterialData.Has("Tables")) {
          tables = MaterialData["Tables"];
        } else {
          KRATOS_INFO("Read materials") << "No tables defined for material ID: " << rProperty.Id()
                                        << std::endl;
        }
        auto p_constitutive_law = KratosComponents<ConstitutiveLaw>::Get(constitutive_law_name)
                                      .Create(cl_parameters, variables, tables, rProperty);
   ```
- turn `ReadMaterialsUtility::AssignVariablesToProperty` and `ReadMaterialsUtility::AssignTablesToProperty` into free functions to be able to call them in the default Create method.

Please find attached two patches:

- kratos-core-patch.tar.bz2 are my proposed changes to be done in `master`
- kratos-mgis-patch.tar.bz2 are my proposed changes to be done in the `mfront/application` branch

[kratos-patches.zip](https://github.com/KratosMultiphysics/Kratos/files/3988622/kratos-patches.zip)

RiccardoRossi commented:
the problem is that, as a resume of our latest meeting, the  @technical-committe is not in favor of further enriching the Create method API by passing in more parameters (this was stated very explicitly, and i share the point), nor of particulatizing the main read materials.

this leaves open the option of having a mgis_read_material and moves the discussion to how to call such a function from within the existing solvers.





that

RiccardoRossi commented:
btw, 

the main conceptual problem here is that Create is NOT in charge of modifying the properties! note that we pass it in as const!

so tables and variables should be read from the read function which CAN modify the propery to store them the tables, etc.

there are at a very minimum 2 reasons behind this choice:
1 - CLs shoukd NEVER modify properties (otherwise for example an element coukd modify a property shared by many other elements!, and even worst, could do it from within a parallel loop!!
2 - ensure the interface is consistent across kratos.

i understand that 2 is limiting the freedom...it is kind of designed for it... we can sodestep by a different reading function

other than that, we can propose a big overhaul of the design, but it will take a lot if discussions before being accepted so it cannot be a shirt term solution

thelfer commented:
@RiccardoRossi, @loumalouomega I understand the design. Let the tables aside, it probably needs more discussion.

Concerning properties, maybe could we mix ideas from @loumalouomega' proposal and mine as follows:

- Pass the variables as `const Parameters` to the `Create` method (and not the properties). The default one should check that the variables are registered and throws if not (current behavior). The `Create` method would allow registered variables and variables expected by the CLs.
- Use @loumalouomega' versions of the `ReadMaterialsUtility::AssignVariablesToProperty` which fills `rProperties` and discard every variables not registered.

Would it be satisfying ?

loumalouomega committed:
Merge branch 'master' into core/read-properties-indepently

loumalouomega commented:
> @RiccardoRossi, @loumalouomega I understand the design. Let the tables aside, it probably needs more discussion.
> 
> Concerning properties, maybe could we mix ideas from @loumalouomega' proposal and mine as follows:
> 
>     * Pass the variables as `const Parameters` to the `Create` method (and not the properties). The default one should check that the variables are registered and throws if not (current behavior). The `Create` method would allow registered variables and variables expected by the CLs.
> 
>     * Use @loumalouomega' versions of the `ReadMaterialsUtility::AssignVariablesToProperty` which fills `rProperties` and discard every variables not registered.
> 
> 
> Would it be satisfying ?

I will wait for the response of @RiccardoRossi

loumalouomega committed:
Merge branch 'master' into core/read-properties-indepently

thelfer commented:
@RiccardoRossi up ?

RiccardoRossi commented:
ouch i forgot completely about this. SORRY.

i'll try to get back to speed asap.



On Thu, Jan 9, 2020 at 12:51 PM thelfer <notifications@github.com> wrote:

> @RiccardoRossi <https://github.com/RiccardoRossi> up ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/KratosMultiphysics/Kratos/pull/6136?email_source=notifications&email_token=AB5PWELDSDYP33CIK4U2INDQ44FUXA5CNFSM4J5KFIK2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEIQBLHQ#issuecomment-572528030>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB5PWEILR62LR4WOJ36OJHDQ44FUXANCNFSM4J5KFIKQ>
> .
>


-- 


*Riccardo Rossi*

PhD, Civil Engineer


member of the Kratos Team: www.cimne.com/kratos

Associate Professor at Universitat Politècnica de Catalunya, BarcelonaTech
(UPC)

Full Research Professor at International Center for Numerical Methods in
Engineering (CIMNE)


C/ Gran Capità, s/n, Campus Nord UPC, Building C1, First Floor

08034 – Barcelona – Spain – www.cimne.com  -

T.(+34) 93 401 56 96 skype: *rougered4*



<http://www.cimne.com/>

<https://www.facebook.com/cimne> <http://blog.cimne.com/>
<http://vimeo.com/cimne> <http://www.youtube.com/user/CIMNEvideos>
<http://www.linkedin.com/company/cimne> <https://twitter.com/cimne>

Les dades personals contingudes en aquest missatge són tractades amb la
finalitat de mantenir el contacte professional entre CIMNE i voste. Podra
exercir els drets d'accés, rectificació, cancel·lació i oposició,
dirigint-se a cimne@cimne.upc.edu. La utilització de la seva adreça de
correu electronic per part de CIMNE queda subjecte a les disposicions de la
Llei 34/2002, de Serveis de la Societat de la Informació i el Comerç
Electronic.

 Imprimiu aquest missatge, només si és estrictament necessari.
<http://www.cimne.com/>

RiccardoRossi commented:
So, essentially my proposal is to make a workaroudn to enable what vicente was proposing first by making it a little cleaner.
i am involving also @philbucher for discussion

right now we have in the structural_mechanics_solver 
~~~py
    def import_constitutive_laws(self):
        materials_filename = self.settings["material_import_settings"]["materials_filename"].GetString()
        if (materials_filename != ""):
            # Add constitutive laws and material properties from json file to model parts.
            material_settings = KratosMultiphysics.Parameters("""{"Parameters": {"materials_filename": ""}} """)
            material_settings["Parameters"]["materials_filename"].SetString(materials_filename)
            KratosMultiphysics.ReadMaterialsUtility(material_settings, self.model)
            materials_imported = True
        else:
            materials_imported = False
        return materials_imported
~~~

We could modify this to 

~~~py
    def import_constitutive_laws(self):
        materials_filename = self.settings["material_import_settings"]["materials_filename"].GetString()

        if(self.settings["material_import_settings"].HasValue("custom_reader")): ##here we use our own file for reading
               custom_reader = import(self.settings["material_import_settings"] "custom_reader").GetString)
               custom_reader.read(self.settings["material_import_settings"])
               materials_imported = True
        else #follow the normal path
            if (materials_filename != ""):
                # Add constitutive laws and material properties from json file to model parts.
               material_settings = KratosMultiphysics.Parameters("""{"Parameters": {"materials_filename": ""}} """)
               material_settings["Parameters"]["materials_filename"].SetString(materials_filename)
               KratosMultiphysics.ReadMaterialsUtility(material_settings, self.model)
               materials_imported = True
           else:
               materials_imported = False
        return materials_imported
~~~

the custom_reader can now do what you like, including calling ReadMaterialsUtility internally.

RiccardoRossi commented:
btw @philbucher @loumalouomega i don't understand why you are creating a new parameter here

~~~py
# Add constitutive laws and material properties from json file to model parts.
            material_settings = KratosMultiphysics.Parameters("""{"Parameters": {"materials_filename": ""}} """)
            material_settings["Parameters"]["materials_filename"].SetString(materials_filename)
~~~

loumalouomega commented:
> ```python
> tring(materials_filename)
> ```



> 
> 
> btw @philbucher @loumalouomega i don't understand why you are creating a new parameter here
> 
> ```python
> # Add constitutive laws and material properties from json file to model parts.
>             material_settings = KratosMultiphysics.Parameters("""{"Parameters": {"materials_filename": ""}} """)
>             material_settings["Parameters"]["materials_filename"].SetString(materials_filename)
> ```

An error was raised, I don-t remember the details right now, and this avoided this error

loumalouomega committed:
Merge branch 'master' into core/read-properties-indepently

philbucher commented:
> So, essentially my proposal is to make a workaroudn to enable what vicente was proposing first by making it a little cleaner.
> i am involving also @philbucher for discussion
> 
> right now we have in the structural_mechanics_solver
> 
> ```python
>     def import_constitutive_laws(self):
>         materials_filename = self.settings["material_import_settings"]["materials_filename"].GetString()
>         if (materials_filename != ""):
>             # Add constitutive laws and material properties from json file to model parts.
>             material_settings = KratosMultiphysics.Parameters("""{"Parameters": {"materials_filename": ""}} """)
>             material_settings["Parameters"]["materials_filename"].SetString(materials_filename)
>             KratosMultiphysics.ReadMaterialsUtility(material_settings, self.model)
>             materials_imported = True
>         else:
>             materials_imported = False
>         return materials_imported
> ```
> 
> We could modify this to
> 
> ```python
>     def import_constitutive_laws(self):
>         materials_filename = self.settings["material_import_settings"]["materials_filename"].GetString()
> 
>         if(self.settings["material_import_settings"].HasValue("custom_reader")): ##here we use our own file for reading
>                custom_reader = import(self.settings["material_import_settings"] "custom_reader").GetString)
>                custom_reader.read(self.settings["material_import_settings"])
>                materials_imported = True
>         else #follow the normal path
>             if (materials_filename != ""):
>                 # Add constitutive laws and material properties from json file to model parts.
>                material_settings = KratosMultiphysics.Parameters("""{"Parameters": {"materials_filename": ""}} """)
>                material_settings["Parameters"]["materials_filename"].SetString(materials_filename)
>                KratosMultiphysics.ReadMaterialsUtility(material_settings, self.model)
>                materials_imported = True
>            else:
>                materials_imported = False
>         return materials_imported
> ```
> 
> the custom_reader can now do what you like, including calling ReadMaterialsUtility internally.

this looks good to me, it is a clean and flexible solution :+1:

RiccardoRossi commented:
i think that we can give for accepted my proposal. Vicente will you have time to make the modificaitons needed?

loumalouomega commented:
> 
> 
> i think that we can give for accepted my proposal. Vicente will you have time to make the modificaitons needed?

Can you summarize the modifications?

RiccardoRossi commented:
> So, essentially my proposal is to make a workaroudn to enable what vicente was proposing first by making it a little cleaner.
> i am involving also @philbucher for discussion
> 
> right now we have in the structural_mechanics_solver
> 
> ```python
>     def import_constitutive_laws(self):
>         materials_filename = self.settings["material_import_settings"]["materials_filename"].GetString()
>         if (materials_filename != ""):
>             # Add constitutive laws and material properties from json file to model parts.
>             material_settings = KratosMultiphysics.Parameters("""{"Parameters": {"materials_filename": ""}} """)
>             material_settings["Parameters"]["materials_filename"].SetString(materials_filename)
>             KratosMultiphysics.ReadMaterialsUtility(material_settings, self.model)
>             materials_imported = True
>         else:
>             materials_imported = False
>         return materials_imported
> ```
> 
> We could modify this to
> 
> ```python
>     def import_constitutive_laws(self):
>         materials_filename = self.settings["material_import_settings"]["materials_filename"].GetString()
> 
>         if(self.settings["material_import_settings"].HasValue("custom_reader")): ##here we use our own file for reading
>                custom_reader = import(self.settings["material_import_settings"] "custom_reader").GetString)
>                custom_reader.read(self.settings["material_import_settings"])
>                materials_imported = True
>         else #follow the normal path
>             if (materials_filename != ""):
>                 # Add constitutive laws and material properties from json file to model parts.
>                material_settings = KratosMultiphysics.Parameters("""{"Parameters": {"materials_filename": ""}} """)
>                material_settings["Parameters"]["materials_filename"].SetString(materials_filename)
>                KratosMultiphysics.ReadMaterialsUtility(material_settings, self.model)
>                materials_imported = True
>            else:
>                materials_imported = False
>         return materials_imported
> ```
> 
> the custom_reader can now do what you like, including calling ReadMaterialsUtility internally.

the change just above + your original proposal (custom reader)

loumalouomega commented:
> 
> 
> > So, essentially my proposal is to make a workaroudn to enable what vicente was proposing first by making it a little cleaner.
> > i am involving also @philbucher for discussion
> > right now we have in the structural_mechanics_solver
> > ```python
> >     def import_constitutive_laws(self):
> >         materials_filename = self.settings["material_import_settings"]["materials_filename"].GetString()
> >         if (materials_filename != ""):
> >             # Add constitutive laws and material properties from json file to model parts.
> >             material_settings = KratosMultiphysics.Parameters("""{"Parameters": {"materials_filename": ""}} """)
> >             material_settings["Parameters"]["materials_filename"].SetString(materials_filename)
> >             KratosMultiphysics.ReadMaterialsUtility(material_settings, self.model)
> >             materials_imported = True
> >         else:
> >             materials_imported = False
> >         return materials_imported
> > ```
> > 
> > 
> > We could modify this to
> > ```python
> >     def import_constitutive_laws(self):
> >         materials_filename = self.settings["material_import_settings"]["materials_filename"].GetString()
> > 
> >         if(self.settings["material_import_settings"].HasValue("custom_reader")): ##here we use our own file for reading
> >                custom_reader = import(self.settings["material_import_settings"] "custom_reader").GetString)
> >                custom_reader.read(self.settings["material_import_settings"])
> >                materials_imported = True
> >         else #follow the normal path
> >             if (materials_filename != ""):
> >                 # Add constitutive laws and material properties from json file to model parts.
> >                material_settings = KratosMultiphysics.Parameters("""{"Parameters": {"materials_filename": ""}} """)
> >                material_settings["Parameters"]["materials_filename"].SetString(materials_filename)
> >                KratosMultiphysics.ReadMaterialsUtility(material_settings, self.model)
> >                materials_imported = True
> >            else:
> >                materials_imported = False
> >         return materials_imported
> > ```
> > 
> > 
> > the custom_reader can now do what you like, including calling ReadMaterialsUtility internally.
> 
> the change just above + your original proposal (custom reader)

OK, but I thought this was already done, I see it is not

loumalouomega committed:
Merge branch 'master' into core/read-properties-indepently

loumalouomega committed:
@RiccardoRossi suggestion

loumalouomega commented:
@RiccardoRossi suggestion applied

philbucher commented on kratos/utilities/read_materials_utility.cpp:
```diff
@@ -169,15 +169,15 @@ void ReadMaterialsUtility::AssignMaterialToProperty(
     AssignTablesToProperty(MaterialData, rProperty);
+    
+    // Assign CL
+    AssignConstitutiveLawToProperty(MaterialData, rProperty);
```
is this change still necessary?

Otherwise looks good to me :+1:

loumalouomega replied:
If we want to avoid some potential errors we should

philbucher replied:
what error?
I am hesitant modifying such an important class without a good reason

loumalouomega replied:
Basically there are checks that are trigered when reading the variables if the order of the reading is done in the current manner, changing the order solves the issue and preserves the current behaviour (because right now there isn't a CL trying to deduce any internal variable from not registered variables)

RiccardoRossi replied:
I am favorable to this change, even though i think it is actually not needed any longer

loumalouomega commented:
@RiccardoRossi can you check this?

philbucher commented:
BTW I realized that I have the exact same problem, I also have a different routine for my simulations
I ended up writing a process that does the reading in `Initialize`
Not sure if a better or worse solution

RiccardoRossi commented on kratos/utilities/read_materials_utility.cpp:
```diff
@@ -222,8 +222,36 @@ void ReadMaterialsUtility::AssignVariablesToProperty(
             const Parameters value = variables.GetValue(iter.name());
+            std::string variable_name = iter.name();
+            TrimComponentName(variable_name);
+            // We create an axiliar parameters for adding the variables
```
what are u doing here???

loumalouomega replied:
To avoid an error when reading properties. The current implementation throws an error for non registered variables. We want to read new defined properties, not registered as variables, that the CL will store. As the first comment of this PR says:

"This PR creates an auxiliar Parameters when reading properties in order to avoid error, so these non-registered properties can be processed later"

RiccardoRossi replied:
I still do not follow. If you try to read an unregistered variable that's by default an error.

 if you want to change this behaviour it must be in the custom reader not in the general one

loumalouomega replied:
I did in the past a custom reader and it was rejected, and this was what we agreed. In any case a warning is thrown

RiccardoRossi replied:
the point of this PR should actually exactly be about allowing a user to define __from the input__ a "custom_reader" 
(see the first file you modified).

In my opinion reading an unexisting variable should be an error not a warning (as it is now)
If you want to have a warning instead you can simply do that in the custom reader and specify in the input that you will be reading that.

loumalouomega replied:
We already allowed to derive readers, it was a discussion then, and we said to do these changes. I can revert previous behaviour but keeping the code implemented here

RiccardoRossi replied:
The current behaviour is safer for "standard" reading.
The discussion was that we should not change the mechanical_solver intrusively to do somethign special. Right now we are defining (first file changed in this PR) a way to do a custom reading elegantly, with only a minor change in the input file. 

I really see as a better solution to keep the current (safe) behaviour in the read_materials_utility.py and simply derive another class from it...

loumalouomega replied:
I understand, but we have that conversation, and I think we were completely confused of the final output

loumalouomega replied:
I will fix it

loumalouomega commented on kratos/utilities/read_materials_utility.cpp:
```diff
@@ -222,8 +222,36 @@ void ReadMaterialsUtility::AssignVariablesToProperty(
+            } else if(KratosComponents<Variable<std::string> >::Has(variable_name)) {
+                auxiliar_variables.AddValue(variable_name, value);
+            } else {
+                KRATOS_WARNING("Read materials") << "The variable property: " << variable_name << " for material ID: " << rProperty.Id() << " is not a registered variable. It will be skipped, and assumed that this value will be processed later" << std::endl;
```
See, there is a warning here

loumalouomega committed:
Merge branch 'master' into core/read-properties-indepently

loumalouomega committed:
Refactor to make @RiccardoRossi happy

loumalouomega commented:
@RiccardoRossi I created a virtual method called RequiresFilterFailingVariables, which is false by default. If set to true does the behaviour I poroposed, but by default preserves current behaviour

loumalouomega committed:
I forgot

RiccardoRossi commented on kratos/utilities/read_materials_utility.cpp:
```diff
@@ -222,8 +222,10 @@ void ReadMaterialsUtility::AssignVariablesToProperty(
         Parameters variables = MaterialData["Variables"];
-        for (auto iter = variables.begin(); iter != variables.end(); ++iter) {
-            const Parameters value = variables.GetValue(iter.name());
+        const Parameters variables_considered = FilterFailingVariablesIfRequired(variables, rProperty.Id());
```
Ok, now it is more of my liking,

however what about making this function virtual (maybe called more genarally FilterVariables)

and implementing it directly in the derived class (the base class would do nothing)? that would be more flexible

loumalouomega replied:
I thought about that, but the problem is the following. The implementation was already here, and I could not move the implementation to the MFront application because is not public yet. With the proposed one I only need to implement a true returning method, instead of the code I added and I would need to move in order to preserve this implementation

RiccardoRossi replied:
An even better would be to essentially make virtual  all the functions in the ReadMaterialUtility (and adding an "empty" FilterVariables in the base class)

Even if i understand that you do not have access to the MFront app, you could simply pass the Filtering code to @thelfer (or if you want to do it yourself, adding a custom reader, IN A DIFFERENT FILE, which is what needs to be done anyway).
That way you don't need to modify the base class (which is used by everyone)

RiccardoRossi replied:
one last petition and i'll approve: can we call this function "FilterVariables" ? or something less specfic than the name you just picked?

loumalouomega replied:
Okay

loumalouomega committed:
Adding virtual

loumalouomega committed:
Merge branch 'master' into core/read-properties-indepently

loumalouomega commented:
~~~cpp
Parameters ModifiedReadMaterialsUtility::FilterFailingVariablesIfRequired(
    const Parameters VariablesParameters,
    const IndexType PropertyId
    )
{
    KRATOS_TRY;
    
    Parameters variables_considered;
    for (auto iter = VariablesParameters.begin(); iter != VariablesParameters.end(); ++iter) {
        const Parameters value = VariablesParameters.GetValue(iter.name());
        std::string variable_name = iter.name();
        TrimComponentName(variable_name);
        // We create an axiliar parameters for adding the variables
        if (KratosComponents<Variable<double> >::Has(variable_name)) {
            variables_considered.AddValue(variable_name, value);
        } else if(KratosComponents<Variable<bool> >::Has(variable_name)) {
            variables_considered.AddValue(variable_name, value);
        } else if(KratosComponents<Variable<int> >::Has(variable_name)) {
            variables_considered.AddValue(variable_name, value);
        } else if(KratosComponents<Variable<array_1d<double, 3> > >::Has(variable_name)) {
            variables_considered.AddValue(variable_name, value);
        } else if(KratosComponents<Variable<array_1d<double, 6> > >::Has(variable_name)) {
            variables_considered.AddValue(variable_name, value);
        } else if(KratosComponents<Variable<Vector > >::Has(variable_name)) {
            variables_considered.AddValue(variable_name, value);
        } else if(KratosComponents<Variable<Matrix> >::Has(variable_name)) {
            variables_considered.AddValue(variable_name, value);
        } else if(KratosComponents<Variable<std::string> >::Has(variable_name)) {
            variables_considered.AddValue(variable_name, value);
        } else {
            KRATOS_WARNING("Read materials") << "The variable property: " << variable_name << " for material ID: " << PropertyId << " is not a registered variable. It will be skipped, and assumed that this value will be processed later" << std::endl;
        }
    }
    return variables_considered;
    
    KRATOS_CATCH("");
}
~~~

The code to be added in the derived class

loumalouomega committed:
Remove code

loumalouomega commented:
> 
> 
> ```c++
> Parameters ModifiedReadMaterialsUtility::FilterFailingVariablesIfRequired(
>     const Parameters VariablesParameters,
>     const IndexType PropertyId
>     )
> {
>     KRATOS_TRY;
>     
>     Parameters variables_considered;
>     for (auto iter = VariablesParameters.begin(); iter != VariablesParameters.end(); ++iter) {
>         const Parameters value = VariablesParameters.GetValue(iter.name());
>         std::string variable_name = iter.name();
>         TrimComponentName(variable_name);
>         // We create an axiliar parameters for adding the variables
>         if (KratosComponents<Variable<double> >::Has(variable_name)) {
>             variables_considered.AddValue(variable_name, value);
>         } else if(KratosComponents<Variable<bool> >::Has(variable_name)) {
>             variables_considered.AddValue(variable_name, value);
>         } else if(KratosComponents<Variable<int> >::Has(variable_name)) {
>             variables_considered.AddValue(variable_name, value);
>         } else if(KratosComponents<Variable<array_1d<double, 3> > >::Has(variable_name)) {
>             variables_considered.AddValue(variable_name, value);
>         } else if(KratosComponents<Variable<array_1d<double, 6> > >::Has(variable_name)) {
>             variables_considered.AddValue(variable_name, value);
>         } else if(KratosComponents<Variable<Vector > >::Has(variable_name)) {
>             variables_considered.AddValue(variable_name, value);
>         } else if(KratosComponents<Variable<Matrix> >::Has(variable_name)) {
>             variables_considered.AddValue(variable_name, value);
>         } else if(KratosComponents<Variable<std::string> >::Has(variable_name)) {
>             variables_considered.AddValue(variable_name, value);
>         } else {
>             KRATOS_WARNING("Read materials") << "The variable property: " << variable_name << " for material ID: " << PropertyId << " is not a registered variable. It will be skipped, and assumed that this value will be processed later" << std::endl;
>         }
>     }
>     return variables_considered;
>     
>     KRATOS_CATCH("");
> }
> ```
> 
> The code to be added in the derived class

BTW, I need this: https://github.com/KratosMultiphysics/Kratos/pull/5997

philbucher commented:
> BTW, I need this: #5997

@loumalouomega could you please integrate this change here such that the required modifications are in one place to review them?

philbucher commented on applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py:
```diff
@@ -311,15 +314,20 @@ def get_mechanical_solution_strategy(self):
-            KratosMultiphysics.ReadMaterialsUtility(material_settings, self.model)
+        if self.settings["material_import_settings"].Has("custom_reader"): # We use our own file for reading
+            custom_reader = import_module(self.settings["material_import_settings"]["custom_reader"].GetString())
+            custom_reader.read(self.settings["material_import_settings"])
```
```suggestion
            custom_reader.ReadMaterials(self.model, self.settings["material_import_settings"])
```

you forgot to pass the `Model`

loumalouomega commented:
> 
> 
> > BTW, I need this: #5997
> 
> @loumalouomega could you please integrate this change here such that the required modifications are in one place to review them?

The change is not for this PR, is for the @RiccardoRossi proposal. BTW #5997 is  very minor change (moving to protected)

loumalouomega committed:
Suggestion

Co-Authored-By: Philipp Bucher <philipp.bucher@tum.de>

loumalouomega commented:
Ping @RiccardoRossi

RiccardoRossi commented:
did u see my last comment about the name?

loumalouomega commented:
> 
> 
> did u see my last comment about the name?

No, did you send it via email?

loumalouomega commented:
> 
> 
> did u see my last comment about the name?

Can you point the comment directly?

RiccardoRossi commented:
I am asking to give a generic name to the new function. On the line of "FilterVariables"

loumalouomega committed:
Merge branch 'master' into core/read-properties-indepently

loumalouomega committed:
FilterVariables

loumalouomega commented:
> FilterVariables

Done

RiccardoRossi reviewed (approved):
ok, i am good with how it is now

loumalouomega commented:
@thelfer now we need to update the MFront branch

loumalouomega commented:
@thelfer I applied changes suggested in the MGIS branch

thelfer commented:
@loumalouomega Thanks ! I'll check this during the week.
